File: secure_cookie.js
#!/usr/bin/env node
'use strict';

/**
 * Secure session cookie encryption using AES-256-GCM (Node.js crypto).
 * This prevents padding oracle attacks by ensuring integrity and authenticity.
 */

const crypto = require('crypto');

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 12;  // 96-bit nonce
const TAG_LENGTH = 16; // 128-bit authentication tag

/**
 * Encrypt and authenticate data.
 * @param {Buffer} data - Data to encrypt.
 * @param {Buffer} key - 32-byte key.
 * @returns {string} Base64-encoded nonce + ciphertext + tag.
 */
function encryptCookie(data, key) {
    if (key.length !== 32) {
        throw new Error('Key must be 32 bytes');
    }
    const iv = crypto.randomBytes(IV_LENGTH);
    const cipher = crypto.createCipheriv(ALGORITHM, key, iv, { authTagLength: TAG_LENGTH });
    let encrypted = cipher.update(data);
    encrypted = Buffer.concat([encrypted, cipher.final()]);
    const tag = cipher.getAuthTag();
    // Combine: iv + encrypted + tag
    const combined = Buffer.concat([iv, encrypted, tag]);
    return combined.toString('base64');
}

/**
 * Decrypt and verify authentication.
 * @param {string} encrypted - Base64-encoded ciphertext.
 * @param {Buffer} key - 32-byte key.
 * @returns {Buffer} Decrypted data.
 * @throws {Error} If authentication fails or format invalid.
 */
function decryptCookie(encrypted, key) {
    if (key.length !== 32) {
        throw new Error('Key must be 32 bytes');
    }
    const combined = Buffer.from(encrypted, 'base64');
    if (combined.length < IV_LENGTH + TAG_LENGTH) {
        throw new Error('Invalid ciphertext length');
    }
    const iv = combined.slice(0, IV_LENGTH);
    const tag = combined.slice(-TAG_LENGTH);
    const ciphertext = combined.slice(IV_LENGTH, -TAG_LENGTH);
    const decipher = crypto.createDecipheriv(ALGORITHM, key, iv, { authTagLength: TAG_LENGTH });
    decipher.setAuthTag(tag);
    try {
        let decrypted = decipher.update(ciphertext);
        decrypted = Buffer.concat([decrypted, decipher.final()]);
        return decrypted;
    } catch (err) {
        // In production, log error and do not reveal padding details
        throw new Error('Decryption failed: cookie may be tampered');
    }
}

// Example usage (for testing) - do not hardcode keys in production
if (require.main === module) {
    const key = crypto.randomBytes(32);
    const cookieData = Buffer.from(JSON.stringify({ user: 'admin', role: 'user' }));
    const enc = encryptCookie(cookieData, key);
    console.log('Encrypted cookie:', enc);
    const dec = decryptCookie(enc, key);
    console.log('Decrypted:', dec.toString('utf8'));
    // Attempt tampering (should fail)
    try {
        const tampered = enc.slice(0, -1) + (enc.slice(-1) === 'a' ? 'b' : 'a');
        decryptCookie(tampered, key);
    } catch (e) {
        console.log('Tampered cookie detected:', e.message);
    }
}

module.exports = { encryptCookie, decryptCookie };

File: secure_cookie.py
#!/usr/bin/env python3
"""
Secure session cookie encryption using AES-GCM (authenticated encryption).
This prevents padding oracle attacks by ensuring integrity and authenticity.
"""
import os
import base64
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.backends import default_backend

def encrypt_cookie(data: bytes, key: bytes) -> str:
    """
    Encrypt and authenticate data using AES-256-GCM.
    Returns base64-encoded nonce + ciphertext + tag.
    """
    if len(key) != 32:  # AES-256 requires 32-byte key
        raise ValueError("Key must be 32 bytes")
    nonce = os.urandom(12)  # 96-bit nonce for GCM
    cipher = Cipher(algorithms.AES(key), modes.GCM(nonce), backend=default_backend())
    encryptor = cipher.encryptor()
    ciphertext = encryptor.update(data) + encryptor.finalize()
    # Return combined: nonce + ciphertext + tag
    combined = nonce + ciphertext + encryptor.tag
    return base64.b64encode(combined).decode('utf-8')

def decrypt_cookie(encrypted: str, key: bytes) -> bytes:
    """
    Decrypt and verify authentication.
    Raises exception if tampered.
    """
    if len(key) != 32:
        raise ValueError("Key must be 32 bytes")
    combined = base64.b64decode(encrypted)
    if len(combined) < 28:  # nonce(12) + tag(16) minimum
        raise ValueError("Invalid ciphertext")
    nonce = combined[:12]
    tag = combined[-16:]
    ciphertext = combined[12:-16]
    cipher = Cipher(algorithms.AES(key), modes.GCM(nonce, tag), backend=default_backend())
    decryptor = cipher.decryptor()
    try:
        return decryptor.update(ciphertext) + decryptor.finalize()
    except Exception:
        # In production, log error and do not reveal padding info
        raise ValueError("Decryption failed: cookie may be tampered")

# Example usage (for testing) - do not hardcode keys in production
if __name__ == '__main__':
    key = os.urandom(32)
    cookie_data = b'{"user":"admin","role":"user"}'
    enc = encrypt_cookie(cookie_data, key)
    print("Encrypted cookie:", enc)
    dec = decrypt_cookie(enc, key)
    print("Decrypted:", dec.decode('utf-8'))
    # Attempt tampering (should fail)
    try:
        tampered = enc[:-1] + ('A' if enc[-1] != 'A' else 'B')
        decrypt_cookie(tampered, key)
    except ValueError as e:
        print("Tampered cookie detected:", e)
